magic_the_gathering/phases/base.py

```diff
@@ -0,0 +1,19 @@
+from abc import abstractmethod
+
+from magic_the_gathering.game_state import GameState
+
+
+class Phase:
+    def __init__(
+        self,
+        name: str,
+        allows_sorcery_speed_when_stack_is_empty: bool = False,
+        players_get_priority: bool = True,
+    ):
+        self.name = name
+        self.allows_sorcery_speed_when_stack_is_empty = allows_sorcery_speed_when_stack_is_empty
+        self.players_get_priority = players_get_priority
+
+    @abstractmethod
+    def run(self, game_state: GameState) -> GameState:
```

Comment:
To be though about more, but I think we'll need to have a `on_entering` and `on_leaving` methods: ex: on the draw phase the draw happens on entering the phase, but on the end step the discarding if hand too big happens on leaving it
I guess this could also be parts of the `run`

Reply:
Ok I see, in fact I had doubts about having only one `run` method. We could split it into 3 methods which by default do nothing (and when it is appropriate we can override them) :slightly_smiling_face: